drivers/gpio/gpio_mchp_xec.c

```diff
@@ -70,64 +85,132 @@ static int gpio_xec_configure(struct device *dev,
+			} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+				*gpio_base &= ~BIT(pin);
+			}
+			pcr1 |= MCHP_GPIO_CTRL_DIR_OUTPUT;
 		}
 	} else { /* GPIO_ACCESS_BY_PORT is not supported */
-		return -EINVAL;
+		return -ENOTSUP;
 	}
 
 	/* Figure out the pullup/pulldown configuration and keep it in the
 	 * pcr1 variable
 	 */
 	mask |= MCHP_GPIO_CTRL_PUD_MASK;
 
-	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+	if ((flags & GPIO_PULL_UP) != 0U) {
 		/* Enable the pull and select the pullup resistor. */
 		pcr1 |= MCHP_GPIO_CTRL_PUD_PU;
-	} else if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+	} else if ((flags & GPIO_PULL_DOWN) != 0U) {
 		/* Enable the pull and select the pulldown resistor */
 		pcr1 |= MCHP_GPIO_CTRL_PUD_PD;
 	}
 
+	/* Push-pull or open drain */
+	mask |= MCHP_GPIO_CTRL_BUFT_MASK;
+
+	if ((flags & GPIO_OPEN_DRAIN) != 0U) {
+		/* Open drain */
+		pcr1 |= MCHP_GPIO_CTRL_BUFT_OPENDRAIN;
+	} else {
+		/* Push-pull */
+		pcr1 |= MCHP_GPIO_CTRL_BUFT_PUSHPULL;
+	}
+
+	/* Use Gpio output register for writing in order to have simetry with
+	 * respect of Gpio input
+	 */
+	mask |= MCHP_GPIO_CTRL_AOD_MASK;
+	pcr1 |= MCHP_GPIO_CTRL_AOD_MASK;
+
+	/* If pin needs to be inverted */
+	mask |= MCHP_GPIO_CTRL_POL_MASK;
+
+	if ((flags & GPIO_ACTIVE_LOW) != 0U) {
+		/* Active Low */
+		pcr1 |= MCHP_GPIO_CTRL_POL_INVERT;
+	} else {
+		/* Active High */
+		pcr1 |= MCHP_GPIO_CTRL_POL_NON_INVERT;
+	}
+
+	/* Now write contents of pcr1 variable to the PCR1 register that
+	 * corresponds to the GPIO being configured
+	 */
+	current_pcr1 = config->pcr1_base + pin;
+	*current_pcr1 = (*current_pcr1 & ~mask) | pcr1;
+
+	return 0;
+}
+
+static int gpio_xec_pin_interrupt_configure(struct device *dev,
+		unsigned int pin, enum gpio_int_mode mode,
+		enum gpio_int_trig trig)
+{
+	const struct gpio_xec_config *config = dev->config->config_info;
+	__IO u32_t *current_pcr1;
+	u32_t pcr1 = 0U;
+	u32_t mask = 0U;
+	u32_t gpio_interrupt = 0U;
+
+	/* Validate pin number range in terms of current port */
+	if ((valid_ctrl_masks[config->port_num] & BIT(pin)) == 0) {
+		return -EINVAL;
+	}
+
+	/* Check if GPIO port supports interrupts */
+	if ((mode != GPIO_INT_MODE_DISABLED) &&
+	    ((config->flags & GPIO_INT_ENABLE) == 0U)) {
+		return -ENOTSUP;
+	}
+
 	/* Assemble mask for level/edge triggered interrrupts */
 	mask |= MCHP_GPIO_CTRL_IDET_MASK;
 
-	if (flags & GPIO_INT) {
-		if (flags & GPIO_INT_EDGE) {
-			/* Enable edge interrupts */
-			if (flags & GPIO_INT_ACTIVE_HIGH) {
-				gpio_interrupt = MCHP_GPIO_CTRL_IDET_REDGE;
-			} else if (flags & GPIO_INT_DOUBLE_EDGE) {
-				gpio_interrupt = MCHP_GPIO_CTRL_IDET_BEDGE;
-			} else {
-				gpio_interrupt = MCHP_GPIO_CTRL_IDET_FEDGE;
-			}
-		} else { /* GPIO_INT_LEVEL */
-			if (flags & GPIO_INT_ACTIVE_HIGH) {
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		/* Explicitly disable interrupts, otherwise the configuration
+		 * results in level triggered/low interrupts
+		 */
+		pcr1 |= MCHP_GPIO_CTRL_IDET_DISABLE;
```

Comment:
A call to enable followed by a call to disable leaves the pin enabled in the GIRQ. Previous code was also doing this.

Reply:
Added a line to clear the interrupt enable bit a few lines above.